dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs

```diff
@@ -1026,15 +1058,30 @@ private static ChatRequestMessage GetRequestMessage(ChatMessageContent message)
                 }
             }
 
-            if (tools is not null)
+            // Handling function calls supplied via ChatMessageContent.Items collection elements of the FunctionCallContent type.
+            if (message.Items.OfType<FunctionCallContent>() is { } functionCallContents && functionCallContents.Any())
             {
-                foreach (ChatCompletionsToolCall tool in tools)
+                var ftcs = new List<ChatCompletionsToolCall>(tools ?? Enumerable.Empty<ChatCompletionsToolCall>());
+
+                foreach (var fcContent in functionCallContents)
                 {
-                    asstMessage.ToolCalls.Add(tool);
+                    if (!ftcs.Any(ftc => ftc.Id == fcContent.Id))
```

Comment:
This is O(N^2).

Separate from complexity, why is this check even needed? In what situation do we end up with two twoo calls with the same ID and we don't want to invoke it that many times?

Reply:
We need both function-calling models (not LLM models) - the current one that uses connector-specific classes and the new one that uses a connector-agnostic approach - to work alongside each other so that existing code continues to function when support for the new model is released. Eventually we may consider deprecating the connector-specific one.

Because of the two models exists alongside each other, there are cases when chat history has assistant messages with two duplicating pieces of information about the same call. 

The first one belongs to the current function calling model and represented either by the OpenAIChatMessageContent.ToolCall property or OpenAIChatMessageContent.FunctionToolCallsProperty metadata. 

The other one belongs to the new function calling model and represented by an item of the FunctionCallRequestContent type.

Reply:
Agreed offline to attempt removing code that reads tool calls from `OpenAIChatMessageContent.ToolCalls` and from `OpenAIChatMessageContent.Metadata` collections in favor of the new model. 

Additionally, we agreed to examine the impact of this change on consumer code that may deserialize a persisted chat history serialized using the current function-call model (not the one added by the PR) that relies on tool call information saved in chat message metadata with the `OpenAIChatMessageContent.FunctionToolCallsProperty` key.

Reply:
Results:
1. So, if the code that reads tool calls from `OpenAIChatMessageContent.ToolCalls` and from `OpenAIChatMessageContent.Metadata` collections is removed, no tool calls (tools LLM requested to invoke) are returned back to the LLM, and it responds with the error: _"400 (Bad Request) - Invalid parameter: messages with the role 'tool' must be a response to a preceding message with 'tool_calls'."_
2. If, in addition to not returning tool calls to LLM, the tool invocation results are not provided either, the LLM re-invokes the same tools/functions even though the chat history contains the final answer that was initially generated by LLM by calling the tools.

Reply:
> This is O(N^2).
> Optimized and simplified.